File: ruby/lib/supersendtx/resources.rb
# frozen_string_literal: true

module SuperSendTX
  class EmailsResource
    def initialize(http)
      @http = http
    end

    def list(limit: nil, cursor: nil)
      @http.request("GET", "/emails#{@http.query({ "limit" => limit, "cursor" => cursor })}")
    end

    def get(email_id)
      @http.request("GET", "/emails/#{URI.encode_www_form_component(email_id)}")
    end

    def send(from: nil, to: nil, **params)
      params["from"] = from if from
      params["to"] = to if to
      body = serialize_send_params(params)
      headers = {}
      idempotency_key = params["idempotency_key"] || params["idempotencyKey"]
      headers["Idempotency-Key"] = idempotency_key.to_s if idempotency_key

      @http.request("POST", "/emails", body: body, headers: headers)
    end

    def batch(emails)
      serialized = emails.map { |email| serialize_send_params(email) }
      @http.request("POST", "/emails/batch", body: { "emails" => serialized })
    end

    def cancel(email_id)
      @http.request("PATCH", "/emails/#{URI.encode_www_form_component(email_id)}", body: { "cancel" => true })
    end

    def resend(email_id)
      @http.request("POST", "/emails/#{URI.encode_www_form_component(email_id)}/resend")
    end

    def test_webhook(**params)
      @http.request("POST", "/emails/test", body: params)
    end

    def insights(window: "30d")
      @http.request("GET", "/deliverability#{@http.query({ "window" => window })}")
    end

    private

    def serialize_send_params(params)
      body = {
        "from" => params["from"] || params[:from],
        "to" => params["to"] || params[:to]
      }

      %w[subject html text reply_to replyTo cc bcc tags headers tag template].each do |key|
        value = params[key] || params[key.to_sym]
        next if value.nil?

        mapped = key == "replyTo" ? "reply_to" : key
        body[mapped] = value
      end

      body["html"] = params["htmlBody"] || params[:htmlBody] if params["htmlBody"] || params[:htmlBody]
      body["text"] = params["textBody"] || params[:textBody] if params["textBody"] || params[:textBody]
      body["scheduled_at"] = params["scheduled_at"] || params[:scheduled_at] if params["scheduled_at"] || params[:scheduled_at]
      body["scheduled_at"] ||= params["scheduledAt"] || params[:scheduledAt]
      body["unsubscribe"] = params["unsubscribe"] || params[:unsubscribe] if params.key?("unsubscribe") || params.key?(:unsubscribe)

      body
    end
  end

  class DomainsResource
    def initialize(http)
      @http = http
    end

    def list(limit: nil, cursor: nil, inbound_enabled: nil)
      @http.request(
        "GET",
        "/domains#{@http.query({ "limit" => limit, "cursor" => cursor, "inbound_enabled" => inbound_enabled })}"
      )
    end

    def get(id_or_name)
      @http.request("GET", "/domains/#{URI.encode_www_form_component(id_or_name)}")
    end

    def create(name, inbound_enabled: nil)
      body = { "name" => name }
      body["inbound_enabled"] = inbound_enabled unless inbound_enabled.nil?
      @http.request("POST", "/domains", body: body)
    end

    def verify(id_or_name)
      @http.request("POST", "/domains/#{URI.encode_www_form_component(id_or_name)}", body: { "action" => "verify" })
    end

    def apply(id_or_name, provider: "cloudflare", credentials: nil)
      body = { "action" => "apply", "provider" => provider }
      body["credentials"] = credentials if credentials
      @http.request("POST", "/domains/#{URI.encode_www_form_component(id_or_name)}", body: body)
    end

    def update(id_or_name, **params)
      @http.request("PATCH", "/domains/#{URI.encode_www_form_component(id_or_name)}", body: params)
    end

    def delete(id_or_name)
      @http.request("DELETE", "/domains/#{URI.encode_www_form_component(id_or_name)}")
    end
  end

  class WebhooksResource
    def initialize(http)
      @http = http
    end

    def list(limit: nil, cursor: nil)
      @http.request("GET", "/webhooks#{@http.query({ "limit" => limit, "cursor" => cursor })}")
    end

    def get(webhook_id)
      @http.request("GET", "/webhooks/#{URI.encode_www_form_component(webhook_id)}")
    end

    def create(**params)
      @http.request("POST", "/webhooks", body: params)
    end

    def update(webhook_id, **params)
      @http.request("PATCH", "/webhooks/#{URI.encode_www_form_component(webhook_id)}", body: params)
    end

    def delete(webhook_id)
      @http.request("DELETE", "/webhooks/#{URI.encode_www_form_component(webhook_id)}")
    end
  end

  class TemplatesResource
    def initialize(http)
      @http = http
    end

    def list(limit: nil, cursor: nil, status: nil)
      @http.request(
        "GET",
        "/templates#{@http.query({ "limit" => limit, "cursor" => cursor, "status" => status })}"
      )
    end

    def get(id_or_alias)
      @http.request("GET", "/templates/#{URI.encode_www_form_component(id_or_alias)}")
    end

    def create(**params)
      @http.request("POST", "/templates", body: params)
    end

    def update(id_or_alias, **params)
      @http.request("PATCH", "/templates/#{URI.encode_www_form_component(id_or_alias)}", body: params)
    end

    def delete(id_or_alias)
      @http.request("DELETE", "/templates/#{URI.encode_www_form_component(id_or_alias)}")
    end

    def publish(id_or_alias)
      @http.request("POST", "/templates/#{URI.encode_www_form_component(id_or_alias)}", body: { "action" => "publish" })
    end
  end

  class SuppressionsResource
    def initialize(http)
      @http = http
    end

    def list(limit: nil, cursor: nil, email: nil)
      @http.request(
        "GET",
        "/suppressions#{@http.query({ "limit" => limit, "cursor" => cursor, "email" => email })}"
      )
    end

    def create(**params)
      @http.request("POST", "/suppressions", body: params)
    end

    def remove(id_or_email)
      if id_or_email.include?("@")
        return @http.request("DELETE", "/suppressions#{@http.query({ "email" => id_or_email })}")
      end

      @http.request("DELETE", "/suppressions/#{URI.encode_www_form_component(id_or_email)}")
    end
  end
end

File: python/supersendtx/resources.py
from __future__ import annotations

from typing import Any

from supersendtx.http import HttpClient


class EmailsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(self, *, limit: int | None = None, cursor: str | None = None) -> dict[str, Any]:
        return self._http.request("GET", f"/emails{self._http.query({'limit': limit, 'cursor': cursor})}")

    def get(self, email_id: str) -> dict[str, Any]:
        return self._http.request("GET", f"/emails/{email_id}")

    def send(
        self,
        *,
        from_: str | None = None,
        to: str | None = None,
        **params: Any,
    ) -> dict[str, Any]:
        if from_ is not None:
            params["from"] = from_
        if to is not None:
            params["to"] = to
        body = _serialize_send_params(params)
        headers: dict[str, str] = {}
        idempotency_key = params.get("idempotency_key") or params.get("idempotencyKey")
        if idempotency_key:
            headers["Idempotency-Key"] = str(idempotency_key)
        return self._http.request("POST", "/emails", body=body, headers=headers)

    def batch(self, emails: list[dict[str, Any]]) -> dict[str, Any]:
        serialized = [_serialize_send_params(email) for email in emails]
        return self._http.request("POST", "/emails/batch", body={"emails": serialized})

    def cancel(self, email_id: str) -> dict[str, Any]:
        return self._http.request("PATCH", f"/emails/{email_id}", body={"cancel": True})

    def resend(self, email_id: str) -> dict[str, Any]:
        return self._http.request("POST", f"/emails/{email_id}/resend")

    def test_webhook(self, **params: Any) -> dict[str, Any]:
        return self._http.request("POST", "/emails/test", body=params)

    def insights(self, *, window: str = "30d") -> dict[str, Any]:
        return self._http.request("GET", f"/deliverability{self._http.query({'window': window})}")


class DomainsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        inbound_enabled: bool | None = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "GET",
            f"/domains{self._http.query({'limit': limit, 'cursor': cursor, 'inbound_enabled': inbound_enabled})}",
        )

    def get(self, id_or_name: str) -> dict[str, Any]:
        return self._http.request("GET", f"/domains/{id_or_name}")

    def create(self, name: str, *, inbound_enabled: bool | None = None) -> dict[str, Any]:
        body: dict[str, Any] = {"name": name}
        if inbound_enabled is not None:
            body["inbound_enabled"] = inbound_enabled
        return self._http.request("POST", "/domains", body=body)

    def verify(self, id_or_name: str) -> dict[str, Any]:
        return self._http.request("POST", f"/domains/{id_or_name}", body={"action": "verify"})

    def apply(
        self,
        id_or_name: str,
        *,
        provider: str = "cloudflare",
        credentials: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        body: dict[str, Any] = {"action": "apply", "provider": provider}
        if credentials:
            body["credentials"] = credentials
        return self._http.request("POST", f"/domains/{id_or_name}", body=body)

    def update(self, id_or_name: str, **params: Any) -> dict[str, Any]:
        return self._http.request("PATCH", f"/domains/{id_or_name}", body=params)

    def delete(self, id_or_name: str) -> dict[str, Any]:
        return self._http.request("DELETE", f"/domains/{id_or_name}")


class WebhooksResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(self, *, limit: int | None = None, cursor: str | None = None) -> dict[str, Any]:
        return self._http.request(
            "GET",
            f"/webhooks{self._http.query({'limit': limit, 'cursor': cursor})}",
        )

    def get(self, webhook_id: str) -> dict[str, Any]:
        return self._http.request("GET", f"/webhooks/{webhook_id}")

    def create(self, **params: Any) -> dict[str, Any]:
        return self._http.request("POST", "/webhooks", body=params)

    def update(self, webhook_id: str, **params: Any) -> dict[str, Any]:
        return self._http.request("PATCH", f"/webhooks/{webhook_id}", body=params)

    def delete(self, webhook_id: str) -> dict[str, Any]:
        return self._http.request("DELETE", f"/webhooks/{webhook_id}")


class TemplatesResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        status: str | None = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "GET",
            f"/templates{self._http.query({'limit': limit, 'cursor': cursor, 'status': status})}",
        )

    def get(self, id_or_alias: str) -> dict[str, Any]:
        return self._http.request("GET", f"/templates/{id_or_alias}")

    def create(self, **params: Any) -> dict[str, Any]:
        return self._http.request("POST", "/templates", body=params)

    def update(self, id_or_alias: str, **params: Any) -> dict[str, Any]:
        return self._http.request("PATCH", f"/templates/{id_or_alias}", body=params)

    def delete(self, id_or_alias: str) -> dict[str, Any]:
        return self._http.request("DELETE", f"/templates/{id_or_alias}")

    def publish(self, id_or_alias: str) -> dict[str, Any]:
        return self._http.request("POST", f"/templates/{id_or_alias}", body={"action": "publish"})


class SuppressionsResource:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        limit: int | None = None,
        cursor: str | None = None,
        email: str | None = None,
    ) -> dict[str, Any]:
        return self._http.request(
            "GET",
            f"/suppressions{self._http.query({'limit': limit, 'cursor': cursor, 'email': email})}",
        )

    def create(self, **params: Any) -> dict[str, Any]:
        return self._http.request("POST", "/suppressions", body=params)

    def remove(self, id_or_email: str) -> dict[str, Any]:
        if "@" in id_or_email:
            return self._http.request("DELETE", f"/suppressions{self._http.query({'email': id_or_email})}")
        return self._http.request("DELETE", f"/suppressions/{id_or_email}")


def _serialize_send_params(params: dict[str, Any]) -> dict[str, Any]:
    body: dict[str, Any] = {
        "from": params["from"],
        "to": params["to"],
    }
    for key in (
        "subject",
        "html",
        "text",
        "reply_to",
        "replyTo",
        "cc",
        "bcc",
        "tags",
        "headers",
        "tag",
        "template",
        "attachments",
    ):
        if key in params and params[key] is not None:
            mapped = "reply_to" if key == "replyTo" else key
            body[mapped] = params[key]
    if params.get("htmlBody") is not None:
        body["html"] = params["htmlBody"]
    if params.get("textBody") is not None:
        body["text"] = params["textBody"]
    if params.get("scheduled_at") is not None:
        body["scheduled_at"] = params["scheduled_at"]
    if params.get("scheduledAt") is not None:
        body["scheduled_at"] = params["scheduledAt"]
    if params.get("unsubscribe") is not None:
        body["unsubscribe"] = params["unsubscribe"]
    return body
